lib/Optimizer/TrainingPreparation.cpp

```diff
@@ -0,0 +1,147 @@
+    if (inputIdx == BN->ScaleIdx) {
+      tensor->init(glow::Tensor::InitKind::Zero, 0, F->getPRNG());
+    } else if (inputIdx == BN->BiasIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    } else if (inputIdx == BN->MeanIdx) {
+      tensor->zero();
+    } else if (inputIdx == BN->VarIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 1.0, F->getPRNG());
+    }
+    break;
+  }
+  case Kinded::Kind::FullyConnectedNodeKind: {
+    FullyConnectedNode *FCN = llvm::dyn_cast<FullyConnectedNode>(node);
+    if (inputIdx == FCN->WeightsIdx) {
+      auto in = FCN->getInput();
+      tensor->init(glow::Tensor::InitKind::Zero, in.dims()[1], F->getPRNG());
+    } else if (inputIdx == FCN->BiasIdx) {
+      tensor->init(glow::Tensor::InitKind::Broadcast, 0.1, F->getPRNG());
+    }
+    break;
+  }
+  default:
+    break;
+  }
+}
+} // namespace
+
+TensorInitializer getDefaultTensorInitializer() {
+  return defaultTensorInitializer;
+}
+
+llvm::Error prepareFunctionForTraining(Function *F,
+                                       PlaceholderBindings &bindings,
+                                       TensorInitializer &&initializer) {
+
+  auto &nodes = F->getNodes();
+
+  Node *outputNode{nullptr};
+
+  // Find SaveNode and its input. It can be SoftMax or other aggregation nodes.
+  // Prevent them from being replaced by Placeholders.
+  for (auto &node : nodes) {
+    if (Kinded::Kind::SaveNodeKind == node.getKind()) {
+      SaveNode *SN = llvm::dyn_cast<SaveNode>(&node);
+      if (!SN) {
+        continue;
+      }
+      // Detected SaveNode.
+      ReshapeNode *RSN = llvm::dyn_cast<ReshapeNode>(SN->getInput().getNode());
+      if (!RSN) {
+        continue;
+      }
+      // Found Input ReshapeNode. Assign its input.
+      outputNode = RSN->getInput().getNode();
+      break;
+    }
+  }
+
+  RETURN_ERR_IF_NOT(outputNode, "Cannot find output node");
+  // Lookup all nodes, skip Storage types, enumerate inputs,
+  // replace Constant type with trainable Placeholders except special cases,
+  // like BatchNormalization inputs (mean and variance). In special cases
+  // replace Constant type with non-trainable Placeholders.
+
+  for (auto &node : nodes) {
+    // Skip storages.
+    if (llvm::isa<Storage>(&node)) {
+      continue;
+    }
+
+    // Skip output node.
+    if (outputNode == &node) {
+      continue;
+    }
+
+    for (unsigned idx = 0, e = node.getNumInputs(); idx < e; ++idx) {
+      auto *IN = node.getNthInput(idx).getNode();
+      Constant *C = llvm::dyn_cast<Constant>(IN);
+      if (!C) {
+        continue;
+      }
+      bool isTrainable = true;
+      switch (node.getKind()) {
+        // Special case. Mean and Variance inputs are not trainable.
+      case Kinded::Kind::BatchNormalizationNodeKind: {
+        BatchNormalizationNode *BN =
+            llvm::dyn_cast<BatchNormalizationNode>(&node);
+        if (BN && (idx = BN->MeanIdx || idx == BN->VarIdx)) {
```

Comment:
nit: Similar to earlier comments -- instead of casting in here and needing to check the node plus follow pointers to get the idx from the enum, the last few lines could be replaced with `if (idx == BatchNormalizationNode::MeanIdx && idx == BatchNormalizationNode::VarIdx)`.

Reply:
Also there's a typo here, missing an `=` in `idx = BN->MeanIdx`